### PR TITLE
Add hashes to requirements.txt

### PR DESCRIPTION
**Pull request by leplatrem**
Ref #1517

### CODE REVIEW

**Comment by Natim** [comment]
Do you have a pip command to do so?

**Comment by leplatrem** [comment]
> Do you have a pip command to do so?

I initialized it with `hashin` but now dependabot is going to update them automatically

**Comment by Natim** [comment]
My understanding is that pip doesn't support this feature but pipenv does, does it?

**Comment by leplatrem** [comment]
> My understanding is that pip doesn't support this feature but pipenv does, does it?

pip does! we use it in kinto-dist already

File: requirements.txt [hunk 1]
```diff
@@ -1,45 +1,253 @@
...
+    --hash=sha256:e0b86084f1e2e78c451994410de756deba206884d6bed68d5a3d7f39ff5fea1d \
+    --hash=sha256:eb86520753560a7e89639500e2a254bb6f683342af598088cb72c73edcad21e6 \
+    --hash=sha256:ff18c5c40a38d41811c23e2480615425c97ea81fd7e9118b8b899c512d97c737
+pycparser==2.19 \
+    --hash=sha256:a988718abfad80b6b157acce7bf130a30876d27603738ac39f140993246b25b3
+pyramid==1.10.1 \
+    --hash=sha256:37c3e1c9eae72817e0365e2a38143543aee8b75240701fa5cb3a1be86c01a1c0 \
+    --hash=sha256:3d5a6165854237b41ac67ced7697d8a884ab2b97625fabfa41b9e53d91187698
 pyramid-multiauth==0.9.0
 pyramid-tm==2.2.1
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply requirements.txt]: I don't explain it...

```
$ hashin pyramid-tm
$ hashin pyramid-multiauth
$ git status
$
```

[the remark this answers is not in the document]
Comment (peterbe) [reply requirements.txt]: Lemme investigate if this is a bug in `hashin`.

[the remark this answers is not in the document]
Comment (peterbe) [reply requirements.txt]: Lemme investigate if this is a bug in `hashin`.

**Comment by peterbe** [comment]
@leplatrem How did you make this change? Or rather, I suspect your version of `hashin` is old. 

I checked out `master` and ran this:
```bash
cat ~/kinto/requirements.txt | xargs python hashin.py -r /tmp/reqs.txt
```

It produced this: https://gist.github.com/peterbe/c68d723ba7543b6ff5c51e9833994ff1

Note that the package isn't called `pyramid-tm` but [`pyramid_tm`](https://pypi.org/project/pyramid_tm/) and it seems an older version of `hashin` fails on that whereas the latest version *corrects* that. 

Anyway, instead of obsessing over how or when, I just went ahead and pushed a change too this PR.

**Comment by peterbe** [comment]
Why are `tox` and `wheel` and `setuptools` in `dev-requirements.txt`? Is legacy or a good reason?

**Comment by leplatrem** [comment]
> Why are tox and wheel and setuptools in dev-requirements.txt? Is legacy or a good reason?

I presume that they are necessary to release Kinto and publish on pypi. We would consider as *dev* requirements anything a dev uses ;)

**Comment by Natim** [comment]
@peterbe can you elaborate on https://github.com/Kinto/kinto/pull/1987/commits/f1ab03115a90dcdf4304b1645b5b885e970380a0 ?

Why do we keep a requirements.txt file with all the hashes and versions if we don't use it?

**Comment by peterbe** [comment]
We can tidy up the mention of `wheel`, `setuptools` and `tox` later. `wheel` and `setuptools` are always installed, by default, in any new virtualenv. And `tox` is installed by Travis. 

One thing I [did on `hashin`](https://github.com/peterbe/hashin/blob/8da1d5eaac03ac1d0a870ed68ff79e642d862a3b/setup.py#L31) was to mention these as "dev" requirements in `setup.py`. [It's the minimal you need to bootstrap](https://github.com/peterbe/hashin#local-development). 

That means you can do:

```bash
git clone repourl
cd repo
mkvirtualenv --or-whatever-you-prefer
pip install -e ".[dev]"
tox
```

**Comment by Natim** [comment]
The extra requirements for dev is a nice addition. I guess we didn't do it like that because it doesn't make sense to install them from pypi but we can get rid of a dev-requirements.txt file so I guess it worth it.

**Comment by peterbe** [comment]
Let's zoom out a little... I jumped into this issue because, I guess, I'm interested in usage of `hashin` :)

I think it's a mistake to use hashes. Yes, for projects like `kinto-dist` that tries to put together `kinto`, with some other specific stuff, in a very deterministic way. 

I think we should remove ALL requirements files and move all listings of dependencies in the `setup.py`. Then, things like `pip install -e ".[dev]"` will work perfectly. It'll serve to do manual testing and for doing tox testing in CI.

**Comment by Natim** [comment]
That's already possible, we only use requirements.txt to give a known working set of dependencies that we knew were working well together at release time. When we release we don't pin dependencies. We only pin to document our users about which version of the dependencies they can use.

**Comment by Natim** [comment]
> I should probably drop this PR then, no?

I think if dependabot updates the hashes accordingly you can keep it.

**Comment by peterbe** [comment]
So the `requirements.txt` is only used for CI?? 

Arguably the name is misleading and it's a noisy pain. I much prefer the Python lists we currently have in `setup.py`. 

If there are specific versions of stuff we need/want purely for unit and functional tests we should move that into `setup.py` and add another key/value to the `extra_requires` and use `pip install -e ".[test]"` (or something) in tox/travis.

**Comment by Natim** [comment]
It is not used for the CI only, it is used to build reproductible Docker images

**Comment by peterbe** [comment]
So then let's just close this issue. 

Suppose that we move the list of know-good-requirements-for-shipping-a-docker-image was a Python list inside the `setup.py` then Dependabot's automation wouldn't be able to help us test that Kinto builds with the latest versions of stuff.

**Comment by Natim** [comment]
Exactly that's why we have setup.py with non pined version and requirements.txt with pinned version.

That's also why I liked when we had a makefile rule to refresh the requirements.txt file.

**Pull request closed without merging**